task_manager/task_manager.hh

```diff
@@ -0,0 +1,323 @@
+        }
+
+        void finish_failed(std::exception_ptr ex) noexcept {
+            finish_failed(ex, fmt::format("{}", ex));
+        }
+
+        virtual void abort() = 0;
+    };
+    class module {
+    private:
+        task_manager& _tm;
+        std::string _name;
+        task_map _tasks;
+        seastar::gate _gate;
+    public:
+        module(task_manager& tm, std::string name) : _tm(tm), _name(std::move(name)) {} // TODO Deexie protect from global usages.
+
+        static module_ptr make_module(task_manager& tm, std::string name) {
+            auto m = seastar::make_shared<module>(tm, name);
+            tm.register_module(std::move(name), m);
+            return m;
+        }
+
+        seastar::gate& async_gate() noexcept {
+            return _gate;
+        }
+
+        abort_source& abort_source() noexcept {
+            return _tm.abort_source();
+        }
+
+        std::chrono::seconds get_task_ttl() const {
+            return _tm.get_task_ttl();
+        }
+
+        task_manager::task_map& get_tasks() noexcept {
+            return _tasks;
+        }
+
+        const task_manager::task_map& get_tasks() const noexcept {
+            return _tasks;
+        }
+
+        task_manager::task_map& get_all_tasks() noexcept {
+            return _tm.get_all_tasks();
+        }
+
+        const task_manager::task_map& get_all_tasks() const noexcept {
+            return _tm.get_all_tasks();
+        }
+
+        void register_task(task_ptr task) {
+            _tasks[task->get_status().id] = task;
+            _tm.register_task(task);
+        }
+
+        void unregister_task(task_id id) {
+            if (_tasks.contains(id)) {
+                _tasks.erase(id);
+                _tm.unregister_task(id);
+            }
+        }
+
+        future<> stop() noexcept {
+            return _gate.close().then([this] {
+                _tm.unregister_module(_name);
+            });
+        }
+    private:
+        future<unsigned> get_task_shard(task_id id) const;
+        task_ptr retrieve_task(task_id id);
+        future<task_ptr> wait_task(task_id id);
+        friend future<std::vector<task_id>> task_manager::task::run_on_shards(task_ptr task, module_ptr module);
+    };
+public:
+    task_manager(config cfg, abort_source& as) noexcept
+        : _cfg(std::move(cfg))
+        , _as(as)
+        , _update_task_ttl_action([this] { return update_task_ttl(task_ttl()); })
+        , _task_ttl_observer(_cfg.task_ttl.observe(_update_task_ttl_action.make_observer()))
+    {}
+
+    module_ptr get_module(std::string module) noexcept {
+        auto it = _modules.find(module);
+        if (it != _modules.end()) {
+            return it->second;
+        }
+        return nullptr;
+    }
+
+    const module_ptr get_module(std::string module) const noexcept {
+        auto it = _modules.find(module);
+        if (it != _modules.end()) {
+            return it->second;
+        }
+        return nullptr;
+    }
```

Comment:
consider just exposing the modules map and let the callers use it to search by module name.